[ews.webkit.org] Consider project when linking to PR

JonWBedard opened the pull request:
#### 5ac9a1b72cff75d95200fc58425d3c5aec56e126
<pre>
[ews.webkit.org] Consider project when linking to PR
<a href="https://bugs.webkit.org/show_bug.cgi?id=247265">https://bugs.webkit.org/show_bug.cgi?id=247265</a>
rdar://101760435

Reviewed by David Kilzer.

* Tools/CISupport/ews-app/ews/common/github.py:
(GitHubEWS.generate_comment_text_for_change):

Canonical link: <a href="https://commits.webkit.org/256157@main">https://commits.webkit.org/256157@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f81f0a06e982627f3bcd66960bc5e7a93279e0c3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94924 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4072 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27828 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104530 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164794 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98919 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4157 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32253 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87230 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100451 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100593 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/81505 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/72883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38665 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36497 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/19584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4246 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40422 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/81505 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42398 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38817 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-app/ews/common/github.py:
```diff
@@ -236,7 +237,7 @@ def generate_comment_text_for_change(self, change):
         if change.comment_id == -1:
-            pr_url = GitHub.pr_url(change.pr_id)
+            pr_url = GitHub.pr_url(change.pr_id, repository_url=repository_url)
             folded_comment = u'Starting EWS tests for {}. Live statuses available at the PR page, {}'.format(hash_url, pr_url)
```
Is this change only for the url in this email notification?

If so, we should consider just removing this logic of this custom message of `Starting EWS tests for ...`. It anyways doesn't work most of the time because of race conditions.

JonWBedard replied:
That was the situation that prompted this fix, yes.

I think we move forward and land this, then discuss removing the logic. It's needed to be pedantically correct, and I can see a future where, for example, this code gets mutated to include a PR link in every email.

aj062 replied:
ok, sounds good

ddkilzer commented on Tools/CISupport/ews-app/ews/common/github.py:
```diff
@@ -215,7 +215,8 @@ def generate_updated_pr_description(self, description, ews_comment):
 
     def generate_comment_text_for_change(self, change):
-        hash_url = 'https://github.com/{}/commit/{}'.format(change.pr_project, change.change_id)
+        repository_url = 'https://github.com/{}'.format(change.pr_project)
+        hash_url = '{}/commit/{}'.format(repository_url, change.change_id)
```
Let's hope those string values never need to be URL-encoded.  :)

JonWBedard replied:
They shouldn't...we're dealing with them as these intermediate strings because it's easier, but we could imagine a world where we constructed them at the very last moment. `change_id` is always guaranteed to be a hash, and `pr_project` is ultimately a string we control (also, there are restrictions from GitHub on that string, although I'm not familiar with what they care, they presumably forbid certain characters, like `/`)

ddkilzer reviewed (approved):
r=me (no changes necessary; just a comment for thought)

JonWBedard committed:
[ews.webkit.org] Consider project when linking to PR
https://bugs.webkit.org/show_bug.cgi?id=247265
rdar://101760435

Reviewed by David Kilzer.

* Tools/CISupport/ews-app/ews/common/github.py:
(GitHubEWS.generate_comment_text_for_change):

Canonical link: https://commits.webkit.org/256157@main
